### PR TITLE
EncryptECIES: verify that the public key is a secp256k1 point 

### PR DESCRIPTION
**Pull request by ahsan-javaid**
## Description
This PR fixes the issue to verify that the public key is a secp256k1 point. The `encryptECIES` method in `src/packages/encryption/ec.ts` does not verify that the given public key string encodes a point on the secp256k1 curve. So in the PR i have added a method to validate the public key and added test cases as well. 

For PR closes the issue #784

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] API reference/documentation update
- [ ] Other

## Does this introduce a breaking change?
No
## Are documentation updates required?
No

## Testing information
Steps to test:
- npm run test
- Other way of testing this PR is to use the `encryptECIES` method with some bad pubic key and it will reject the public key by throwing an exception.

## Checklist
- [ ] Code is commented where needed
- [x] Unit test coverage for new or modified code paths
- [x] `npm run test` passes
- [ ] Changelog is updated
- [x] Tag 1 of @yknl or @zone117x for review

### CODE REVIEW

File: packages/encryption/src/ec.ts [hunk 1]
```diff
@@ -101,6 +101,48 @@ function sharedSecretToKeys(sharedSecret: Buffer): { encryptionKey: Buffer; hmac
...
+  return maybe.match(/^[0-9a-f]+$/i) !== null;
+}
+
+/**
+ * @ignore
+ */
+function isValidPublicKey(pub: string): { result: boolean; reason: string | null } {
+  const invalidFormat = {
+    result: false,
+    reason: 'Invalid public key format',
```
Comment (kyranjamie) [inline packages/encryption/src/ec.ts]: Maybe it'd be good to use a reason `enum` here, such as

```ts
enum InvalidPublicKeyReason {
  InvalidFormat,
  IsNotPoint
}
```

This'd make it easier to programatically check which one it is. Atm you'd have to compare strings. The text as it is could still be returned, but perhaps as a different property?

**Review by kyranjamie — APPROVED** [review]

File: packages/encryption/src/ec.ts (hunk 1 above)
Comment (ahsan-javaid) [reply packages/encryption/src/ec.ts]: @kyranjamie Added these changes. Let me know if it's fine or need further improvement.

File: packages/encryption/src/ec.ts [hunk 2]
```diff
@@ -101,6 +109,58 @@ function sharedSecretToKeys(sharedSecret: Buffer): { encryptionKey: Buffer; hmac
...
+  };
+  if (pub.length !== 66 && pub.length !== 130) return invalidFormat;
+
+  const firstByte = pub.slice(0, 2);
+
+  // uncompressed public key
+  if (pub.length === 130 && firstByte !== '04') return invalidFormat;
+
+  // compressed public key
+  if (pub.length === 66 && firstByte !== '02' && firstByte !== '03') return invalidFormat;
```
Comment (zone117x) [inline packages/encryption/src/ec.ts:145]: This looks right to me. Could you add test cases for all these branches? Well known valid keys for each version/prefix, and invalid keys for each?

Comment (ahsan-javaid) [reply packages/encryption/src/ec.ts:145]: @zone117x Test cases added for all these branches.

**Review by zone117x — APPROVED** [review]

**Comment by ahsan-javaid** [comment]
@zone117x Is this ready for merge. cc: @diwakergupta

**Pull request merged**